### ISSUE TITLE
can't run gitlab ci pipeline, FAO Scott.wittenburg?

### ISSUE THREAD

**Issue by varioustoxins** [LLNL/spack#29479]
### Steps to reproduce

I am trying to run the CI framework on gitlab to start testing pip installation of packages with @adamjstewart 

starting with example script from the CI tutorial fails on cloud hosted gitlab

```stages: [generate, build]

variables:
  SPACK_REPO: https://github.com/scottwittenburg/spack.git
  SPACK_REF: pipelines-reproducible-builds

generate-pipeline:
  stage: generate
  tags:
    - docker
  image:
    name: ghcr.io/scottwittenburg/ecpe4s-ubuntu18.04-runner-x86_64:2020-09-01
    entrypoint: [""]
  before_script:
    - git clone ${SPACK_REPO}
    - pushd spack && git checkout ${SPACK_REF} && popd
    - . "./spack/share/spack/setup-env.sh"
  script:
    - spack env activate --without-view .
    - spack -d ci generate
      --artifacts-root "${CI_PROJECT_DIR}/jobs_scratch_dir"
      --output-file "${CI_PROJECT_DIR}/jobs_scratch_dir/pipeline.yml"
  artifacts:
    paths:
      - "${CI_PROJECT_DIR}/jobs_scratch_dir"

build-jobs:
  stage: build
  trigger:
    include:
      - artifact: "jobs_scratch_dir/pipeline.yml"
        job: generate-pipeline
    strategy: depend
```

pipeline runs but fails with message

```. "./spack/share/spack/setup-env.sh"
$ spack env activate --without-view .
==> Error: No such environment: '.'

...


Error: 'spack ci generate' requires an environment
  activate an environment first:
      spack env activate ENV
  or use:
      spack -e ENV ci generate ...
```

using latest from develop in 'share/spack/gitlab/cloud_pipelines'

fails to run at all on commit 

### Error message

as above

### Information on your system

gitlab cloud

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by scottwittenburg** [comment]
@varioustoxins That example project on gitlab.com hasn't been updated in awhile, but was intended to be a small standalone example of an environment repository.  So I'm just curious where you were trying this pipeline?  In your own project on gitlab.com or on a self-hosted gitlab?  You mentioned you were using latest from develop in `share/spack/gitlab/cloud_pipelines`, but the example expects the environment to be in the root of the project.  Additionally, the example illustrates how to run pipelines when cloning spack is required, while the pipelines you pointed to are designed for a mirror/fork of spack itself.  If you can provide a link to the failing pipeline, that will make debugging easier, otherwise, more information will help.

**Comment by locsmith** [comment]
hi @scottwittenburg
so this is me working at trying to make sure the python2 builds continue to work... I am currently working on gitlab as it seemed the easiest place to start but will need to move to another hosts when i reach out of time/credit. I was just using gitlab in the cloud to bootstrap things as it seemed the easiest place to start as I presume they will have setup correct. 

_You mentioned you were using latest from develop in share/spack/gitlab/cloud_pipelines, but the example expects the environment to be in the root of the project._ so i had .gitlab-ci.yml in the root of the repo, do i need it somewhere else or do i need a different path for the spack env activate --without-view .

the repo is at https://gitlab.com/garyt.and.sarahb/spack/-/blob/develop

I think I need to work on a clone fork currently but in the end they will also need to run on commits...

**Comment by scottwittenburg** [comment]
>  I was just using gitlab in the cloud to bootstrap things as it seemed the easiest place to start as I presume they will have setup correct.

One thing you'll need to figure out if you use https://gitlab.com is where the binary packages will be stored after each spec gets built and installed.  The example pipeline you linked above only builds a few very small packages, and thanks to that, it can use gitlab artifacts for the binary mirror.  I don't know how large of artifacts gitlab.com is configured to handle, but I think it's out of your control.  If you have access to an S3 bucket, that's a better option, as you can push binary packages straight there from the pipeline.

> so i had .gitlab-ci.yml in the root of the repo, do i need it somewhere else or do i need a different path for the spack env activate --without-view .

Maybe I'm misreading that statement, but the `.gitlab-ci.yml` doesn't have anything to do with `spack env activate --without-view .`.  The former is the yaml configuration of the gitlab pipeline, the latter is a command to activate a spack environment defined by a `spack.yaml` file.  You can have either file anywhere you like, because you control your gitlab repo and thus can set the location and name of the pipeline configuration file, and because you can change into whatever directory you want before activating your spack environment.

Your repo looks like a mirror of spack, but you also clone a different spack as described in the pipelines section of spack readthedocs.  I probably wouldn't do both of those, but pick one or the other.  If I understand your goal (I might be missing something), I feel like you might want to start with a simple repo containing just a spack environment file (`spack.yaml`) and the gitlab pipeline configuration (`gitlab-ci.yml`), just like in the documentation you cited.  However, I would use a more recent version of spack, and one that you control (push a recent version to your fork of spack, and then use that in your pipelines).  That way you can make sure e.g. the version of spack doesn't change in the middle of one of your pipelines.

In your spack environment is where you specify all the python2 packages you care about (instead of building all the stack spack builds on each PR, which is what your project is trying to do now).

**Comment by varioustoxins** [comment]
> > I was just using gitlab in the cloud to bootstrap things as it seemed the easiest place to start as I presume they will have setup correct.
> 
> One thing you'll need to figure out if you use https://gitlab.com is where the binary packages will be stored after each spec gets built and installed. The example pipeline you linked above only builds a few very small packages, and thanks to that, it can use gitlab artifacts for the binary mirror. I don't know how large of artifacts gitlab.com is configured to handle, but I think it's out of your control. If you have access to an S3 bucket, that's a better option, as you can push binary packages straight there from the pipeline.


I don't think this is a problem I just want to use gitlab to get things bootstrapped, there is no intention to save large amounts of binary on gitlab in the end.... in fact once the pipeline is up and running I will most probably transfer to a local instance....

> 
> > so i had .gitlab-ci.yml in the root of the repo, do i need it somewhere else or do i need a different path for the spack env activate --without-view .
> 
> Maybe I'm misreading that statement, but the `.gitlab-ci.yml` doesn't have anything to do with `spack env activate --without-view .`. The former is the yaml configuration of the gitlab pipeline, the latter is a command to activate a spack environment defined by a `spack.yaml` file. You can have either file anywhere you like, because you control your gitlab repo and thus can set the location and name of the pipeline configuration file, and because you can change into whatever directory you want before activating your spack environment.

So I am trying to follow the spack pipelines tutorial at  https://spack.readthedocs.io/en/latest/pipelines.html (though I have tried some more complete examples which failed in a similar manner...)

so I have the .gitlab-ci.yml at top-level as ?recommended? and it contains 

```
 - spack env activate --without-view .
 ```

in the generate section

the seems to be failing so I presume at the runtime the environment is in the wrong place or the directory is wrong...

output from the run is at https://gitlab.com/garyt.and.sarahb/spack/-/jobs/2315074032

and is publicly visible as is the whole cloned repository if that helps


> 
> Your repo looks like a mirror of spack, 

yes this is true I wanted to work with the latest develop...

but you also clone a different spack as described in the pipelines section of spack readthedocs. I probably wouldn't do both of those, but pick one or the other. 

ok not quite sure where that is, are we basically saying that in ```.gitlab-ci.yml``` I am using ```SPACK_REPO: https://github.com/scottwittenburg/spack.git``` whereas I should be using ```https://gitlab.com/garyt.and.sarahb/spack.git```?

that would make sense..



If I understand your goal (I might be missing something), I feel like you might want to start with a simple repo containing just a spack environment file (`spack.yaml`) and the gitlab pipeline configuration (`gitlab-ci.yml`), just like in the documentation you cited. However, I would use a more recent version of spack, and one that you control (push a recent version to your fork of spack, and then use that in your pipelines). That way you can make sure e.g. the version of spack doesn't change in the middle of one of your pipelines.

that makes sense
> 
> In your spack environment is where you specify all the python2 packages you care about (instead of building all the stack spack builds on each PR, which is what your project is trying to do now).

ok that make sense I think I have tried with two .gitlab-ci.yml one from Adam Stewart which contains a lot of packages to build and the tutorial one which is obviously much simpler....not sure which one you saw, I currently have tutorial spack.yml setup which has only two packages to build ( zlib and bzip2) if I read it right.

**Comment by varioustoxins** [comment]
ok problem solved I had a spack.yml not a spack.yaml...

thoughts 

0. mea culpa that was a runaround for a 1 letter error (never done that before!)
1. error message could be clearer
2. it would be nice if it also accepted a file as an environment
3. maybe spack.yml should be supported as well as spack .yaml as both are common extensions for yaml files

**Comment by varioustoxins** [comment]
Hi Scott more problems:  I am having problems with the mirrors bucket being empty


```
 File "/builds/garyt.and.sarahb/spack/spack/bin/spack", line 98, in <module>
1554    sys.exit(spack.main.main())
1555  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/main.py", line 900, in main
1556    return _main(argv)
1557  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/main.py", line 883, in _main
1558    return _invoke_command(command, parser, args, unknown)
1559  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/main.py", line 553, in _invoke_command
1560    return_val = command(parser, args)
1561  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/cmd/ci.py", line 643, in ci
1562    return args.func(args)
1563  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/cmd/ci.py", line 560, in ci_rebuild
1564    env, job_spec_yaml_path, buildcache_mirror_url, sign_binaries
1565  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/ci.py", line 1457, in push_mirror_contents
1566    raise inst
1567  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/ci.py", line 1438, in push_mirror_contents
1568    _push_mirror_contents(env, specfile_path, sign_binaries, mirror_url)
1569  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/ci.py", line 1433, in _push_mirror_contents
1570    bindist.push(matches, push_url, spec_kwargs, **kwargs)
1571  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/binary_distribution.py", line 1211, in push
1572    _build_tarball(node, push_url, **kwargs)
1573  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/binary_distribution.py", line 1008, in _build_tarball
1574    if web_util.url_exists(remote_spackfile_path):
1575  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/util/web.py", line 232, in url_exists
1576    raise err
1577  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/util/web.py", line 227, in url_exists
1578    s3.get_object(Bucket=url.netloc, Key=url.path.lstrip('/'))
1579  File "/usr/local/lib/python3.6/dist-packages/botocore/client.py", line 337, in _api_call
1580    return self._make_api_call(operation_name, kwargs)
1581  File "/usr/local/lib/python3.6/dist-packages/botocore/client.py", line 656, in _make_api_call
1582    raise error_class(parsed_response, operation_name)
1583botocore.errorfactory.NoSuchBucket: An error occurred (NoSuchBucket) when calling the GetObject operation: The specified bucket does not exist
```

and it appears I need a mirror...

```
==> [2022-04-10-15:35:16.610812] Error: spack ci generate requires an env containing a mirror
237Traceback (most recent call last):
238  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/main.py", line 900, in main
239    return _main(argv)
240  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/main.py", line 883, in _main
241    return _invoke_command(command, parser, args, unknown)
242  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/main.py", line 553, in _invoke_command
243    return_val = command(parser, args)
244  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/cmd/ci.py", line 643, in ci
245    return args.func(args)
246  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/cmd/ci.py", line 143, in ci_generate
247    use_dependencies=use_dependencies, artifacts_root=artifacts_root)
248  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/spack/ci.py", line 562, in generate_gitlab_ci_yaml
249    tty.die('spack ci generate requires an env containing a mirror')
250  File "/builds/garyt.and.sarahb/spack/spack/lib/spack/llnl/util/tty/__init__.py", line 275, in die
251    sys.exit(1)
252SystemExit: 1
```

does this mean the mirror is out of date, or do I need an access token  or out of sync or do I need to generate my own mirror?

mirror was defined as 

mirrors: { "mirror": "s3://spack-public/mirror" }

I had a play around with s3fs and it appears I am lacking credentials even to list the public mirror...?

```python
from fs_s3fs import S3FS
from fs import open_fs
s3fs = open_fs('s3://spack-public/mirror')
s3fs.listdir('/')
```

**Comment by varioustoxins** [comment]
hi Scott so an update, I have the sample script working, my big change was that I changed the mirror that was an s3 bucket and was used to push the results to to a file on the local file system which should disappear once the pipeline is complete? This shouldn't be a problem as I am just trying to prove I have successful builds at the moment... and even in the real cases will want to get to a successful build and then run unit tests and dispose...

mirrors: { "mirror": "file:///builds/garyt.and.sarahb/spack" }

I will of course have to work out how to build the mirrors variable on the fly I guess..., not quite sure how you do that yet...

**Comment by alalazo** [comment]
@varioustoxins @scottwittenburg I'm trying to go through untriaged bug reports and triage them. Can you please summarize what is the status here?[^1]

[^1]: The objective is to have all bugs triaged and showing up in https://github.com/orgs/spack/projects/6)

**Comment by scottwittenburg** [comment]
Sorry, I don't know the status here.

**Comment by alalazo** [comment]
Given the amount of changes in pipelines since this issue was opened, I'm going to close it as as outdated. @varioustoxins Feel free to reopen, or to start a new issue, in case you think there's something still to be fixed / discussed.
